src/hex.ts

```diff
@@ -82,27 +82,42 @@ export function isValidHexAddress(possibleAddress: Hex) {
 
 /**
  * Encode a passed hex string as an ERC-55 mixed-case checksum address.
+ * This is the unmemoized version, primarily used for testing.
  *
  * @param address - The hex address to encode.
  * @returns The address encoded according to ERC-55.
  * @see https://eips.ethereum.org/EIPS/eip-55
  */
-export function getChecksumAddress(address: Hex): Hex {
-  assert(is(address, HexChecksumAddressStruct), 'Invalid hex address.');
-  const unPrefixed = remove0x(address.toLowerCase());
-  const unPrefixedHash = remove0x(bytesToHex(keccak256(unPrefixed)));
-  return `0x${unPrefixed
-    .split('')
-    .map((character, nibbleIndex) => {
-      const hashCharacter = unPrefixedHash[nibbleIndex];
-      assert(is(hashCharacter, string()), 'Hash shorter than address.');
-      return parseInt(hashCharacter, 16) > 7
-        ? character.toUpperCase()
-        : character;
-    })
-    .join('')}`;
+export function getChecksumAddressUnmemoized(hexAddress: Hex): Hex {
+  assert(is(hexAddress, HexChecksumAddressStruct), 'Invalid hex address.');
+  const address = remove0x(hexAddress).toLowerCase()
+
+  const hashBytes = keccak256(address)
+  const length = address.length
+  const result = new Array(length) // Pre-allocate array
+
+  for (let i = 0; i < length; i++) {
+    const byteIndex = i >> 1 // Faster than Math.floor(i / 2)
+    const nibbleIndex = i & 1 // Faster than i % 2
+    const byte = hashBytes[byteIndex]!
+    const nibble = nibbleIndex === 0 ? (byte >> 4) : (byte & 0x0f)
+    
+    result[i] = nibble >= 8 ? address[i]!.toUpperCase() : address[i]!
+  }
+
+  return `0x${result.join('')}`
 }
 
+/**
+ * Encode a passed hex string as an ERC-55 mixed-case checksum address.
+ * This function is memoized for performance.
+ *
+ * @param address - The hex address to encode.
+ * @returns The address encoded according to ERC-55.
+ * @see https://eips.ethereum.org/EIPS/eip-55
```

Comment:
Open to discussion - I think it would be great to also have the specification code inline/near the bitwise operations to help readability.

Reply:
I left some comments on code that's explaining what it is doing, I can probably replace it with normal code, final difference is very small.